libbeat/registry/backend/memlog/tx.go

```diff
@@ -0,0 +1,414 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package memlog
+
+import (
+	"github.com/elastic/beats/libbeat/registry/backend"
+)
+
+type transaction struct {
+	store *store
+	state txState
+	mem   memTx
+
+	typeConv *typeConv
+
+	active   bool
+	readonly bool
+}
+
+func newTransaction(store *store, readonly bool) *transaction {
+	tx := &transaction{
+		store:    store,
+		readonly: readonly,
+		active:   true,
+	}
+
+	tx.state.init()
+	tx.mem.init(tx, tx.store.mem, &tx.state)
+	return tx
+}
+
+func (tx *transaction) close() {
+	if tx.active {
+		lock := chooseTxLock(&tx.store.lock, tx.readonly)
+		lock.Unlock()
+		tx.active = false
+
+		if tx.typeConv != nil {
+			tx.typeConv.release()
+			tx.typeConv = nil
+		}
+	}
+}
+
+func (tx *transaction) Close() error {
+	if tx.active {
+		tx.Rollback()
+	}
+	return nil
+}
+
+func (tx *transaction) Rollback() error {
+	if !tx.active {
+		return errTxInactive
+	}
+
+	defer tx.close()
+	tx.mem.Rollback()
```

Comment:
same here

Reply:
`tx.mem.Rollback` has no error return, as all state is in memory. The `(*transaction) Rollback()` method has an error return, so to satisfy the backends interface.